Initial import

ttddyy opened the pull request:
Here is the import from datasource-proxy-r2dbc.

- Copy and paste from datasource-proxy-r2dbc
- Modified package structure
- Added license header
- Incorporated `Wrapped` interface (in separate commit)

ttddyy commented:
Hmm, javadoc generation is failing. Will fix them.

gregturn commented on src/main/java/io/r2dbc/proxy/callback/ProxyFactory.java:
```diff
@@ -0,0 +1,44 @@
+    Statement<?> createStatement(Statement<?> statement, String query, ConnectionInfo connectionInfo);
+
+    Result createResult(Result result, QueryExecutionInfo executionInfo);
+
```
Some of these methods sound strange. `ConnectionFactory createConnectionFactory(ConnectionFactory connectionFactory)` for example. `createConnectionFactory` implies it's creating a new one, but it has one as an input. Does this wrap one with another?

ttddyy replied:
Yes, the return value is the wrapped(proxied) input value.
Probably, rename each method to `createProxy...` for better naming??

Example usage would be:
```java
Connection proxiedConnecition = proxyFactory.createProxyConnection(originalConnection, connectionInfo);
```

gregturn replied:
Much better.

ttddyy replied:
Renamed to `createProxy...`

gregturn commented on src/main/java/io/r2dbc/proxy/callback/ProxyUtils.java:
```diff
@@ -0,0 +1,63 @@
+            return ((ConnectionHolder) result).getOriginalConnection();
+        }
+        return null;
+    }
```
Is this `null` approach the right approach? Does wrapping with `Optional` cost too much?

ttddyy replied:
Good Idea. I'll convert to Optional

gregturn replied:
Check first. In some scenarios Optional adds a bit of overhead. They actually backed off doing it in Spring Data Commons.

ttddyy replied:
These method are utility for users and more for corner case use to retrieve original `Connection` object.
Not in the hot path. So, converted to `Optional`.

mp911de replied:
I also think non-hot code paths are fine for `Optional<…>` usage.

gregturn commented on src/main/java/io/r2dbc/proxy/callback/ReactiveBatchCallback.java:
```diff
@@ -0,0 +1,89 @@
+    private ConnectionInfo connectionInfo;
+
+    private List<String> queries = new ArrayList<>();
+
```
Should this initialization be moved into the constructor call?

ttddyy replied:
Um, yes, it can, but isn't it more common to initialize collection at declaration if they are not assigned by constructor?

gregturn commented on src/main/java/io/r2dbc/proxy/callback/ReactiveConnectionCallback.java:
```diff
@@ -0,0 +1,89 @@
+            String query = (String) args[0];
+            return this.proxyConfig.getProxyFactory().createStatement((Statement) result, query, this.connectionInfo);
+        }
+
```
Is keying off every API's method name via String the only way to manage this?

ttddyy replied:
Another way would be to compare `Method` object.

However, I think there are drawbacks for `Method` comparison.
- Need to programmatically prepare `Method` objects.
- Handling of overloading methods (same name but different args)

So, even though comparing with String method name ties to literal method name typing, it is much easier to use for controlling logic flow.

mp911de replied:
Using strings is quite a common pattern and the `Method` approach is significantly more difficult, as someone might redeclare a method and then it's a different one than retrieved from the interface.

gregturn commented on src/main/java/io/r2dbc/proxy/callback/ReactiveConnectionFactoryCallback.java:
```diff
@@ -0,0 +1,77 @@
+        }
+
+        BiFunction<Object, MethodExecutionInfo, Object> onMap = null;
+        if ("create".equals(methodName)) {
```
Line break

ttddyy replied:
ok

gregturn commented on src/main/java/io/r2dbc/proxy/listener/CompositeProxyExecutionListener.java:
```diff
@@ -0,0 +1,91 @@
+//    public void setListeners(List<ProxyExecutionListener> listeners) {
+//        this.listeners = listeners;
+//    }
+
```
Remove.

ttddyy replied:
yes.

gregturn commented on src/main/java/io/r2dbc/proxy/listener/ProxyExecutionListener.java:
```diff
@@ -0,0 +1,125 @@
+        } else if (eventType == ProxyEventType.AFTER_METHOD) {
+            afterMethod(executionInfo);
+        }
+    }
```
And what if neither condition is met?

ttddyy replied:
Hm, the calling logic kinds of ensures event type is either `BEFORE_METHOD` or `AFTER_METHOD`.

Actually, this is one of the part I changed the way from datasource-proxy jdbc to r2dbc.

In dsp-jdbc, instead of having this dispatcher method, proxy logic directly calls
`beforeMethod()` or `afterMethod()`. So, probably that might be better style than dispatching in default method like here.

I'll remove this method and directly call before/after methods.

gregturn commented on src/main/java/io/r2dbc/proxy/listener/ProxyExecutionListener.java:
```diff
@@ -0,0 +1,125 @@
+        } else if (eventType == ProxyEventType.AFTER_QUERY) {
+            afterQuery(executionInfo);
+        }
+    }
```
What if neither type is met?

ttddyy replied:
will remove this and call before/after directly

gregturn commented on src/test/java/io/r2dbc/proxy/callback/JdkProxyFactoryTest.java:
```diff
@@ -0,0 +1,113 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.mock;
```
Can we stick with one assertion toolkit?

ttddyy replied:
OK, seems like  r2dbc projects uses assertj for all assertions.
I'll convert to it.

ttddyy replied:
Converted all jupiter `assert[Equals/Same/True/False/Throws]` to AssertJ assertions.
phew~ :)

ttddyy commented on src/main/java/io/r2dbc/proxy/callback/DefaultConnectionIdManager.java:
```diff
@@ -0,0 +1,40 @@
+    public String getId(Connection connection) {
+        String id = String.valueOf(this.idCounter.incrementAndGet());
+        return id;
+    }
```
Yes, definitely

ttddyy replied:
done

ttddyy commented on src/main/java/io/r2dbc/proxy/callback/JdkProxyFactory.java:
```diff
@@ -0,0 +1,151 @@
+
+/**
+ * ProxyFactory using JDK dynamic proxy.
+ *
```
Yes, will do

ttddyy commented on src/main/java/io/r2dbc/proxy/callback/JdkProxyFactory.java:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    public static class ConnectionFactoryInvocationHandler implements InvocationHandler {
+
```
No, they don't need to, I'll change this and others to private.

gregturn replied:
Or consider package private. I’m a fan of default visibility.

ttddyy replied:
changed them to package private

ttddyy commented on src/main/java/io/r2dbc/proxy/callback/ProxyConfig.java:
```diff
@@ -0,0 +1,65 @@
+    //    public ProxyConfig(ProxyFactory proxyFactory) {
+    //        this.proxyFactory = proxyFactory;
+    //    }
+
```
Yup

ttddyy commented on src/main/java/io/r2dbc/proxy/callback/CallbackSupport.java:
```diff
@@ -0,0 +1,290 @@
+
+                executionInfo.setProxyEventType(ProxyEventType.AFTER_METHOD);
+                listener.onMethodExecution(executionInfo);
+            }
```
Other ways I can think of is to create a StopWatch, or make a template for the pattern with Consumer.

Can you expand a little bit about try-resource object?

gregturn replied:
Essentially wrap things in a templated method. Perhaps Consumer or Supplier is a best fit. Try-Resource is something that must eventually be closed like a file, so maybe my comment was a little off.

ttddyy replied:
Tried to template the flow, but since it involves with Flux operations, not quite readable than directly putting logic on each Flux methods.
Instead, just made a simple `StopWatch` utility class, and delegated the `Duration` calculation, which seems cleaner approach.

ttddyy commented on src/main/java/io/r2dbc/proxy/callback/ConnectionIdManager.java:
```diff
@@ -0,0 +1,28 @@
+import io.r2dbc.spi.Connection;
+
+/**
+ * @author Tadaya Tsuyukubo
```
Yes, will do.

ttddyy replied:
Added

ttddyy commented on src/main/java/io/r2dbc/proxy/callback/DefaultConnectionIdManager.java:
```diff
@@ -0,0 +1,40 @@
+
+/**
+ * Copy from datasource-proxy
+ *
```
Oops, missed this. will do.

ttddyy commented:
@nebhale OK, got it. I'll update them.
Meanwhile, I just pushed one small commit to fix package directory structure.

mp911de commented on src/main/java/io/r2dbc/proxy/callback/ProxyConfig.java:
```diff
@@ -0,0 +1,63 @@
+
+    public void setProxyFactory(ProxyFactory proxyFactory) {
+        this.proxyFactory = proxyFactory;
+        this.proxyFactory.setProxyConfig(this);
```
That cycle is fishy. I understand that you need to create proxies and pass on some configuration. How about encapsulating either `ProxyFactory` with a `ProxyFactoryFactory` that then accepts a runtime configuration object? Not sure this is the best approach but right now I lack a better idea.

mp911de commented on src/main/java/io/r2dbc/proxy/callback/ProxyFactory.java:
```diff
@@ -0,0 +1,46 @@
+
+    void setProxyConfig(ProxyConfig proxyConfig);
+
+    ConnectionFactory createProxyConnectionFactory(ConnectionFactory connectionFactory);
```
How about reducing API surface in general by providing a single `<T> wrap(T delegate, Object... args)` method? The routing could be handled within the actual factory.

mp911de commented on src/main/java/io/r2dbc/proxy/callback/ProxyUtils.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    public static Optional<Connection> getOriginalConnection(Connection connection) {
```
How about sticking to `unwrap` terminology? Something like `unwrapConnection`?

mp911de commented on src/main/java/io/r2dbc/proxy/callback/CallbackSupport.java:
```diff
@@ -0,0 +1,284 @@
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public abstract class CallbackSupport {
```
How about naming types in this hierarchy `InvocationHandler` (this class would be named `InvocationHandlerSupport`)? Reflects better the actual purpose.

ttddyy replied:
I'm thinking to rename them to `CallbackHandler`.
So, 
- `CallbackSupport` => `CallbackHandlerSupport`
- `ReactiveStatementCallback` => `StatementCallbackHandler`
- `ReactiveConnectionFactoryCallback` => `ConnectionFactoryCallbackHandler`
- etc.

mp911de commented on src/main/java/io/r2dbc/proxy/core/Binding.java:
```diff
@@ -0,0 +1,32 @@
+ * @see Bindings.IndexBinding
+ * @see Bindings.IdentifierBinding
+ */
+public interface Binding {
```
Binding keeps reappearing across various implementations. That's just something I noticed, don't want to fix that right now.

mp911de commented on src/main/java/io/r2dbc/proxy/core/BindingValue.java:
```diff
@@ -0,0 +1,61 @@
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public interface BindingValue {
```
How about `BoundValue`? Users don't have a chance to change it anyways.

mp911de commented on src/main/java/io/r2dbc/proxy/core/BindingValue.java:
```diff
@@ -0,0 +1,61 @@
+public interface BindingValue {
+
+    Object getValue();
+
```
An `isNull()` method would be a good fit here.

mp911de commented on src/main/java/io/r2dbc/proxy/core/ConnectionInfo.java:
```diff
@@ -0,0 +1,129 @@
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public class ConnectionInfo {
```
Let's turn that into an interface and make the default implementation package-private.

mp911de commented on src/main/java/io/r2dbc/proxy/core/MethodExecutionInfo.java:
```diff
@@ -0,0 +1,148 @@
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public class MethodExecutionInfo {
```
Spring/Zipkin have various levels of exposure regarding this kind of object structures. E.g. `MethodExecutionInfo` (exposing read-only methods) and `MutableMethodExecutionInfo` so things get clear very quickly what the intent is.

mp911de commented on src/main/java/io/r2dbc/proxy/core/QueryExecutionInfo.java:
```diff
@@ -0,0 +1,246 @@
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public class QueryExecutionInfo {
```
See `MethodExecutionInfo` comment.

mp911de commented on src/main/java/io/r2dbc/proxy/support/QueryExecutionInfoFormatter.java:
```diff
@@ -0,0 +1,434 @@
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public class QueryExecutionInfoFormatter implements Function<QueryExecutionInfo, String> {
```
That's a decent approach. Care to make `QueryExecutionInfoFormatter` immutable so it can be used similar to Java 8's Date/Time Formatters.

ttddyy commented on src/main/java/io/r2dbc/proxy/ProxyConnectionFactoryBuilder.java:
```diff
@@ -0,0 +1,133 @@
+        return new ProxyConnectionFactoryBuilder(delegate);
+    }
+
+    public static ProxyConnectionFactoryBuilder create(ConnectionFactory delegate, ProxyConfig proxyConfig) {
```
So, something like this?

```java
public class ProxyConnectionFactory {
  public static ProxyConnectionFactoryBuilder create(ConnectionFactory original) {
    return new ProxyConnectionFactoryBuilder(original);
  }
}

public class ProxyConnectionFactoryBuilder {
  public ConnectionFactory build() {...}
  public ProxyConnectionFactoryBuilder listener(...) {...}
  ... other configuration methods ...
}
```

and, usage is:

```java
ConnectionFactory proxy = 
    ProxyConnectionFactory.create(actualConnectionFactory)
        .listener(...)
        .build();
```

compare to:

```java
// current implementation
ConnectionFactory proxy = 
    ProxyConnectionFactoryBuilder.create(actualConnectionFactory)
        .listener(...)
        .build();
```

ttddyy commented:
ok, I have updated the PR.

Summary of changes:
- Add more javadoc
- Add `Assert` class
  - Nullability check
- Restrict visibility of classes
- Use `wrap`, `unwrap` terminology
- Binding related
  - Rename `BindingValue` to `BoundValue`
  - Use `isNull` and consolidate `NullBindingValue` and `SimpleBindingValue`
- Convert `ConnectionInfo`, `MethodExecutionInfo`, `QueryExecutionInfo` to interface
- `ProxyFactoryFactory` to replace circular reference
- Make `Clock` configurable via `ProxyConfig`
- Make `[Query|Method]ExecutionInfoFormatter` immutable
- Rename `Reactive~Callback` to `~CallbackHandler`
- Refactor `JdkProxyFactory` with added `CallbackHandler` interface

----

Regarding visibility of the classes, I changed many of the classes to package scoped final classes. So, it is pretty much locked down and users cannot see or reuse most of them.

In datasource-proxy, I took the other approach. I made classes very open, so that user can reuse them either by overriding them or composing mix of user's custom classes and existing classes.

I just want to make sure in r2dbc projects, rule of thumbs is to hide as much as possible first.
Then, if user requests opening up a class and it is reasonable to do, then we may widen the scope.

nebhale commented:
> I just want to make sure in r2dbc projects, rule of thumbs is to hide as much as possible first.
Then, if user requests opening up a class and it is reasonable to do, then we may widen the scope.

You're on the right path here.  We consider public APIs to be supported with compatibility guarantees which means the larger the surface area, the less future flexibility we have.  We strongly follow the open/closed principal as well in order to manage the area that bugs could be lurking due to ill-defined contracts.

nebhale commented on src/test/java/io/r2dbc/proxy/support/MethodExecutionInfoFormatterTest.java:
```diff
@@ -0,0 +1,113 @@
+        ConnectionInfo connectionInfo = mock(ConnectionInfo.class);
+        when(connectionInfo.getConnectionId()).thenReturn("ABC");
+
+        MethodExecutionInfo executionInfo = mock(MethodExecutionInfo.class);
```
This isn't going to be a binding requirement, but I'd recommend creating a `test` package with full-fledged Mock types for all (or perhaps most important) interface types.  It's pretty obvious that both users _and_ the project need to create fake versions of these types and rather than requiring users to include a mocking library, and write the behavior, wouldn't it be nicer to say:

```java
MockMethodExecutionInfo executionInfo = MockMethodExecutionInfo.builder()
    .threadId(5L)
    .connectionInfo(MockConnectionInfo.builder()
        .connectionId("ABC")
        .build())
    .executeDuration(Duration.ofMillis(23))
    .method(String.class, "indexOf", int.class)
    .target(100L)
    .build()

nebhale commented on src/test/java/io/r2dbc/proxy/support/QueryExecutionInfoFormatterTest.java:
```diff
@@ -0,0 +1,546 @@
+        when(connectionInfo.getConnectionId()).thenReturn("conn-id");
+
+        // Batch Query
+        QueryExecutionInfo execInfo = mock(QueryExecutionInfo.class);
```
This isn't going to be a binding requirement, but I'd recommend creating a `test` package with full-fledged Mock types for all (or perhaps most important) interface types.  It's pretty obvious that both users _and_ the project need to create fake versions of these types and rather than requiring users to include a mocking library, and write the behavior, wouldn't it be nicer to say:

```java
MockQueryExecutionInfo executionInfo = MockQueryExecutionInfoInfo.builder()
    .threadName("my-thread")
    .threadId(99L)
    .connectionInfo(MockConnectionInfo.builder()
        .connectionId("conn-id")
        .build())
    .success(true)
    .executeDuration(Duration.ofMillis(35))
    .type(ExecutionType.BATCH)
    .batchSize(20)
    .bindingsSize(10)
    .queries("SELECT A", "SELECT B")
    .build()

nebhale commented on src/main/java/io/r2dbc/proxy/listener/LifeCycleListener.java:
```diff
@@ -0,0 +1,253 @@
+    // for ConnectionFactory
+    //
+
+    default void beforeCreateOnConnectionFactory(MethodExecutionInfo methodExecutionInfo) {
```
Need documentation on these methods as well, _especially_ around nullability guarantees for implementors.

nebhale commented on src/main/java/io/r2dbc/proxy/core/QueryInfo.java:
```diff
@@ -0,0 +1,46 @@
+
+    private final List<Bindings> bindingsList = new ArrayList<>();
+
+    public QueryInfo(String query) {
```
Needs documentation.
